File: set_03/kaffi/kaffi.py
import math


def can_fit_with_max_height(max_height, chair_counts, w):
    stacks_needed = 0

    print(f"\nChecking max_height: {max_height}")
    print(chair_counts)

    for count in chair_counts:
        stacks_needed += (count + max_height - 1) // max_height
        stacks_needed += math.ceil(count / max_height)

        print(
            f"Count: {count}, Max Height: {max_height}, Stacks Needed: {stacks_needed}"
        )
        if stacks_needed > w:
            print("Exceeded width limit")
            return False

    return stacks_needed <= w


def solve_chair_stacking(w, chair_counts):
    total_chairs = sum(chair_counts)

    left, right = 1, max(chair_counts)

    while left < right:
        mid = (left + right) // 2
        if can_fit_with_max_height(mid, chair_counts, w):
            right = mid
        else:
            left = mid + 1

    max_height = left
    disorganization = w * max_height - total_chairs

    return disorganization


import sys


def main():
    lines = sys.stdin.read().strip().split("\n")
    n, w = map(int, lines[0].split())
    chair_counts = list(map(int, lines[1].split()))

    result = solve_chair_stacking(w, chair_counts)
    print(result)


if __name__ == "__main__":
    main()

File: set_03/kaffi/kaffi.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

inline bool canFitWithMaxHeight(const vector<int> &chairCounts, int w, int maxHeight)
{
    int stacksNeeded = 0;
    for (int count : chairCounts)
    {
        stacksNeeded += (count + maxHeight - 1) / maxHeight;
        if (stacksNeeded > w)
            return false;
    }
    return true;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, w;
    cin >> n >> w;

    vector<int> chairCounts(n);
    int totalChairs = 0;
    int maxCount = 0;

    for (int i = 0; i < n; i++)
    {
        cin >> chairCounts[i];
        totalChairs += chairCounts[i];
        maxCount = max(maxCount, chairCounts[i]);
    }

    int left = 1, right = maxCount;

    while (left < right)
    {
        int mid = left + (right - left) / 2;
        if (canFitWithMaxHeight(chairCounts, w, mid))
        {
            right = mid;
        }
        else
        {
            left = mid + 1;
        }
    }

    int maxHeight = left;
    int disorganization = w * maxHeight - totalChairs;

    cout << disorganization << '\n';

    return 0;
}